File: leetcode/algorithms/tree/n-ary-tree-preorder-traversal.java
/*
    leetcode
    category: "Algorithms"
    problem title: "N-ary Tree Preorder Traversal"
*/


/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    
    private void addToList(Node node,List<Integer> preOrderList){
        if(node==null){
            return;
        }
        preOrderList.add(node.val);
        for(Node child : node.children){
            addToList(child,preOrderList);
        }
    }
    
    public List<Integer> preorder(Node root) {
        List<Integer> preOrderList = new ArrayList<Integer>();
        
        addToList(root,preOrderList);
        
        return preOrderList;
        
    }
}

File: leetcode/algorithms/tree/n-ary-tree-preorder-traversal.js
/*
    leetcode
    category: "Algorithms"
    problem title: "N-ary Tree Preorder Traversal"
*/


/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {Node} root
 * @return {number[]}
 */

const populatePreorder = (root,preorder) => {
    if(root===null){
        return;
    }
    preorder.push(root.val);
    root.children.forEach(child => {
        populatePreorder(child,preorder);
        
    })
    
}


var preorder = function(root) {
    
    const preorder = [];
    
    populatePreorder(root,preorder);
    
    return preorder;
    
    
    
};